### PR TITLE
feat: add birthday integration

### LINKED ISSUE #1339: request: birthday commands
**Describe the solution you'd like**

We've had multiple requests for birthday related commands. They come down to 2 requests:

1. A command to wish a member a happy birthday with some random message: `s!birthday <member>`
2. A command to set reminders for whenever it is someone's birthday, which also automatically schedules a new reminder for the following year. Initially this can be done with user reminders but ideally it would be set up with guild reminders and require a "birthdays" channel where Skyra outputs the reminder to.

### PR DESCRIPTION
**Pull request by Stitch07**
TODOs:

* Better Date parsing
* Birthday GIFs
* Setting timezones

Closes #1339

### CODE REVIEW

File: src/lib/database/database.config.ts [hunk 1]
```diff
@@ -20,7 +20,8 @@ export const config: ConnectionOptions = {
 		migrationsDir: 'src/lib/database/migrations',
 		subscribersDir: 'src/lib/database/subscribers'
 	},
-	namingStrategy: new SnakeNamingStrategy()
+	namingStrategy: new SnakeNamingStrategy(),
+	logging: true
```
Comment (kyranet) [inline src/lib/database/database.config.ts]: Make this a configuration key in config.ts, please.

File: src/lib/i18n/languageKeys/keys/Settings.ts [hunk 2]
```diff
@@ -4,6 +4,10 @@ export const Prefix = T<string>('settings:prefix');
 export const Language = T<string>('settings:language');
 export const DisabledCommands = T<string>('settings:disabledCommands');
 export const DisableNaturalPrefix = T<string>('settings:disableNaturalPrefix');
+// TODOstitch): complete these
```
Comment (kyranet) [inline src/lib/i18n/languageKeys/keys/Settings.ts]: ?

File: src/tasks/birthday.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
+import { getHandler } from '#languages';
+import { PartialResponseValue, ResponseType, Task } from '#lib/database';
+import { TextChannel, User } from 'discord.js';
+
+const enum Matches {
+	Age = '{age}',
+	AgeOrdinal = '{ageOrdinal}',
+	User = '{user}'
```
Comment (kyranet) [inline src/tasks/birthday.ts]: This isn't enough featureful and doesn't have consistency [with tags](https://gist.github.com/kyranet/485287383d818ef70fe347e45f4f63a5).

`{user}` should mention, `{user.name}` should use the username, `{user.tag}` should show the tag, and so on.

Also, `{age.ordinal}` would look more consistent with the others this way.

File: src/tasks/birthday.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+export class UserTask extends Task {
+	private kTransformMessageRegExp = /{age}|{ageOrdinal}|{user}/g;
```
Comment (kyranet) [inline src/tasks/birthday.ts]:
```suggestion
	private kTransformMessageRegExp = /{(?:age|age\.ordinal|user|user\.name|user\.tag|...}/g;
```

File: src/tasks/birthday.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+	public async run(data: BirthdayTaskData): Promise<PartialResponseValue | null> {
+		const guild = this.context.client.guilds.cache.get(data.guildID);
+		if (!guild) return null;
+		const member = await guild.members.fetch(data.userID);
+		if (!member) return null;
+
+		const t = await guild.fetchT();
+		const [birthdayRole, birthdayChannel, birthdayMessage] = await guild.readSettings(['birthdayRole', 'birthdayChannel', 'birthdayMessage']);
+		if (!birthdayRole && !(birthdayChannel && birthdayMessage)) return null;
+		const { me } = guild;
```
Comment (kyranet) [inline src/tasks/birthday.ts]: Never going to be null:
```suggestion
		const me = guild.me!;
```

File: src/tasks/removeBirthdayRole.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
+import { PartialResponseValue, ResponseType, Task } from '#lib/database';
+
+export class UserTask extends Task {
+	public async run(data: RemoveBirthdayRoleData): Promise<PartialResponseValue | null> {
+		const guild = this.context.client.guilds.cache.get(data.guildID);
+		if (!guild) return null;
+		const member = await guild.members.fetch(data.userID);
+		if (!member) return null;
+
+		const { me } = guild;
```
Comment (kyranet) [inline src/tasks/removeBirthdayRole.ts]: Never going to be null:
```suggestion
		const me = guild.me!;
```

File: src/tasks/removeBirthdayRole.ts [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+		if (me?.permissions.has('MANAGE_ROLES') && me?.roles.highest.position > member.roles.highest.position) {
```
Comment (kyranet) [inline src/tasks/removeBirthdayRole.ts]:
```suggestion
		if (me.permissions.has('MANAGE_ROLES') && me.roles.highest.position > member.roles.highest.position) {
```

File: src/tasks/removeBirthdayRole.ts [hunk 8]
```diff
@@ -0,0 +1,23 @@
+import { PartialResponseValue, ResponseType, Task } from '#lib/database';
+
+export class UserTask extends Task {
+	public async run(data: RemoveBirthdayRoleData): Promise<PartialResponseValue | null> {
+		const guild = this.context.client.guilds.cache.get(data.guildID);
+		if (!guild) return null;
+		const member = await guild.members.fetch(data.userID);
```
Comment (kyranet) [inline src/tasks/removeBirthdayRole.ts:8]:
```suggestion

		const member = await guild.members.fetch(data.userID);
```

File: src/tasks/birthday.ts [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+				}
+			});
+		}
+
+		if (birthdayChannel && birthdayMessage) {
+			const channel = guild.channels.cache.get(birthdayChannel);
+			if (!channel || !channel.permissionsFor(me!)?.has('SEND_MESSAGES')) return null;
+			await (channel as TextChannel).send(this.transformMessage(birthdayMessage, member.user, data.birthDate, t.lng));
+		}
+		// next year's birthday
```
Comment (kyranet) [inline src/tasks/birthday.ts]:
```suggestion

		// next year's birthday
```

File: src/tasks/birthday.ts [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+		const nextBirthday = new Date();
+		nextBirthday.setUTCFullYear(nextBirthday.getFullYear() + 1);
+		nextBirthday.setUTCHours(0);
+		nextBirthday.setUTCMinutes(0);
+		nextBirthday.setUTCSeconds(0);
+		nextBirthday.setUTCMilliseconds(0);
```
Comment (kyranet) [inline src/tasks/birthday.ts]: This should be moved into its own method.

File: src/tasks/birthday.ts [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+			await member.roles.add(birthdayRole);
+			const tomorrow = new Date();
+			tomorrow.setDate(tomorrow.getDate() + 1);
+			await this.context.client.schedules.add('removeBirthdayRole', tomorrow, {
+				data: {
+					guildID: data.guildID,
+					roleID: birthdayRole,
+					userID: data.userID
+				}
+			});
```
Comment (kyranet) [inline src/tasks/birthday.ts]: This should be moved into its own method.

File: src/tasks/birthday.ts [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+		if (!guild) return null;
+		const member = await guild.members.fetch(data.userID);
+		if (!member) return null;
+
+		const t = await guild.fetchT();
+		const [birthdayRole, birthdayChannel, birthdayMessage] = await guild.readSettings(['birthdayRole', 'birthdayChannel', 'birthdayMessage']);
+		if (!birthdayRole && !(birthdayChannel && birthdayMessage)) return null;
+		const { me } = guild;
+
+		if (birthdayRole && me?.permissions.has('MANAGE_ROLES') && me?.roles.highest.position > member.roles.highest.position) {
```
Comment (kyranet) [inline src/tasks/birthday.ts]:
```suggestion
		if (birthdayRole && me.permissions.has('MANAGE_ROLES') && me.roles.highest.position > member.roles.highest.position) {
```

File: src/tasks/birthday.ts [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+	User = '{user}'
+}
+
+export class UserTask extends Task {
+	private kTransformMessageRegExp = /{age}|{ageOrdinal}|{user}/g;
+
+	public async run(data: BirthdayTaskData): Promise<PartialResponseValue | null> {
+		const guild = this.context.client.guilds.cache.get(data.guildID);
+		if (!guild) return null;
+		const member = await guild.members.fetch(data.userID);
```
Comment (kyranet) [inline src/tasks/birthday.ts:21]:
```suggestion

		const member = await guild.members.fetch(data.userID);
```

File: src/lib/util/util.ts [hunk 14]
```diff
@@ -641,6 +641,15 @@ export const random = (num: number) => Math.round(Math.random() * num);
 export const sendLoadingMessage = (message: GuildMessage | Message, t: TFunction): Promise<typeof message> =>
 	message.send(new MessageEmbed().setDescription(pickRandom(t(LanguageKeys.System.Loading))).setColor(BrandingColors.Secondary));
 
+export function nextBirthday(birthDate: Date) {
```
Comment (kyranet) [inline src/lib/util/util.ts]: Needs unit tests

File: src/languages/en-US/commands/misc.json [hunk 15]
```diff
@@ -374,7 +412,7 @@
 		"examples": [
 			"HE COMES"
 		],
-		"reminder": "H̡̰͉̜̰̠̣͆ͣ̒ͯ̽̃E̪̱̬̪ͣ̓̈́̐͗́̑ͩ͠ͅ ̞͍͍̪̟͚̩̃̀ͯ̓ͥͪͦ̎͜C̣̮̄ͩͤͬͥ͆͛͆͢͞O̲͙̟͎̙͑̿̋̚M̛̖͙͖̣͇̊ͦ̽ͭ͋̑̂̾Ē̸̳͎͖̟̿̆S̱̭̩̳̠̜ͨ̔̿ͅ"
+		"reminder": "H̡̰͉̜̰̠̣͆ͣ̒ͯ̽̃E̪̱̬̪ͣ̓̈́̐͗́̑ͩ͠ͅ ̞͍͍̪̟͚̩̃̀ͯ̓ͥͪͦ̎͜C̣̮̄ͩͤͬͥ͆͛͆͢͞O̲͙̟͎̙͑̿̋̚M̛̖͙͖̣͇̊ͦ̽ͭ͋̑̂̾Ē̸̳͎͖̟̿̆S̱̭̩̳̠̜ͨ̔̿ͅ"
```
Comment (kyranet) [inline src/languages/en-US/commands/misc.json:420]: Why was this changed?

File: src/commands/Misc/Birthday/view-birthday.ts [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+})
+export default class extends SkyraCommand {
+	public async run(message: Message, args: SkyraCommand.Args) {
+		const user = args.finished ? message.author! : await args.pick('userName');
+		const birthDate = this.context.client.schedules.queue.find(
+			(schedule) => schedule.taskID === 'birthday' && schedule.data.guildID === message.guild!.id && schedule.data.userID === user.id
+		)?.time;
+
+		return message.send(
+			// @ts-ignore i have no idea what causes this
```
Comment (kyranet) [inline src/commands/Misc/Birthday/view-birthday.ts]: This happens because the options in both overloads are different, you need to do

```typescript
birthDate
	? args.t(LanguageKeys.Commands.Misc.ViewBirthdaySet, { birthDate: birthDate.getTime(), user: user.tag })
	: args.t(LanguageKeys.Commands.Misc.ViewBirthdayNotSet, { user: user.tag });
```

File: src/commands/Misc/Birthday/upcoming-birthdays.ts [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+	public async run(message: Message, args: SkyraCommand.Args) {
+		const schedules = Object.entries(
+			this.context.client.schedules.queue
+				.filter((schedule) => schedule.taskID === 'birthday' && schedule.data.guildID === message.guild!.id)
+				.reduce((a, b) => {
+					const key = b.time.getTime();
+					if (Reflect.has(a, key)) Reflect.get(a, key).push(b);
+					else Reflect.set(a, key, [b]);
+					return a;
+				}, {})
```
Comment (kyranet) [inline src/commands/Misc/Birthday/upcoming-birthdays.ts]: Please use a `Map<number, ScheduleEntity[]>`, so properties aren't stringified.

**Review by kyranet — DISMISSED** [review]

File: src/lib/database/database.config.ts (hunk 1 above)
Comment (Stitch07) [reply src/lib/database/database.config.ts]: I'll set it to DEV

File: src/commands/Misc/Birthday/set-birthday.ts [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+	extendedHelp: LanguageKeys.Commands.Misc.SetBirthdayExtended,
+	runIn: ['text']
+})
+export default class extends SkyraCommand {
+	public async run(message: Message, args: SkyraCommand.Args) {
+		const { t } = args;
+		const [birthdayRole, birthdayChannel, birthdayMessage] = await message.guild!.readSettings([
+			'birthdayRole',
+			'birthdayChannel',
+			'birthdayMessage'
```
Comment (favna) [inline src/commands/Misc/Birthday/set-birthday.ts]: Use the `GuildSettings` thingy.

File: src/commands/Misc/Birthday/set-birthday.ts [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+import { Time } from '@sapphire/time-utilities';
+
+@ApplyOptions<SkyraCommand.Options>({
+	cooldown: 10,
+	description: LanguageKeys.Commands.Misc.SetBirthdayDescription,
+	extendedHelp: LanguageKeys.Commands.Misc.SetBirthdayExtended,
+	runIn: ['text']
+})
+export default class extends SkyraCommand {
+	public async run(message: Message, args: SkyraCommand.Args) {
```
Comment (favna) [inline src/commands/Misc/Birthday/set-birthday.ts]:
```suggestion
	public async run(message: GuildMessage, args: SkyraCommand.Args) {
```

File: src/commands/Misc/Birthday/set-birthday.ts [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+		const currentTask = this.context.client.schedules.queue.find(
+			(schedule) => schedule.taskID === 'birthday' && schedule.data.guildID === message.guild!.id && schedule.data.userID === message.author.id
+		);
+		if (currentTask) await this.context.client.schedules.remove(currentTask);
+
+		const birthday = nextBirthday(birthDate);
+		await this.context.client.schedules.add('birthday', birthday, {
+			data: {
+				birthDate,
+				guildID: message.guild!.id,
```
Comment (favna) [inline src/commands/Misc/Birthday/set-birthday.ts]:
```suggestion
				guildID: message.guild.id,
```

File: src/commands/Misc/Birthday/set-birthday.ts [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+		if (!birthdayRole && !(birthdayChannel && birthdayMessage)) return this.error(LanguageKeys.Commands.Misc.SetBirthdayNotConfigured);
+
+		const birthDate = await args.rest('date');
+		birthDate.setTime(birthDate.getTime() - birthDate.getTimezoneOffset() * Time.Minute);
+		// bounds checking
+		// the world's oldest human alive was born in January 1903
+		if (birthDate.getTime() > Date.now() || birthDate.getFullYear() < 1903) return this.error(LanguageKeys.Commands.Misc.SetBirthdayInvalidDate);
+		// delete any existing reminders
+		const currentTask = this.context.client.schedules.queue.find(
+			(schedule) => schedule.taskID === 'birthday' && schedule.data.guildID === message.guild!.id && schedule.data.userID === message.author.id
```
Comment (favna) [inline src/commands/Misc/Birthday/set-birthday.ts]:
```suggestion
			(schedule) => schedule.taskID === 'birthday' && schedule.data.guildID === message.guild.id && schedule.data.userID === message.author.id
```

File: src/commands/Misc/Birthday/set-birthday.ts [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+@ApplyOptions<SkyraCommand.Options>({
+	cooldown: 10,
+	description: LanguageKeys.Commands.Misc.SetBirthdayDescription,
+	extendedHelp: LanguageKeys.Commands.Misc.SetBirthdayExtended,
+	runIn: ['text']
+})
+export default class extends SkyraCommand {
+	public async run(message: Message, args: SkyraCommand.Args) {
+		const { t } = args;
+		const [birthdayRole, birthdayChannel, birthdayMessage] = await message.guild!.readSettings([
```
Comment (favna) [inline src/commands/Misc/Birthday/set-birthday.ts]:
```suggestion
		const [birthdayRole, birthdayChannel, birthdayMessage] = await message.guild.readSettings([
```

File: src/commands/Misc/Birthday/reset-birthday.ts [hunk 23]
```diff
@@ -0,0 +1,21 @@
+import { SkyraCommand } from '#lib/structures';
+import { LanguageKeys } from '#lib/i18n/languageKeys';
+import { ApplyOptions } from '@sapphire/decorators';
+import type { Message } from 'discord.js';
+
+@ApplyOptions<SkyraCommand.Options>({
+	cooldown: 5,
```
Comment (favna) [inline src/commands/Misc/Birthday/reset-birthday.ts]:
```suggestion
	cooldown: 10,
```

File: src/commands/Misc/Birthday/upcoming-birthdays.ts [hunk 24]
```diff
@@ -0,0 +1,46 @@
+import { SkyraCommand } from '#lib/structures';
+import { LanguageKeys } from '#lib/i18n/languageKeys';
+import { ApplyOptions } from '@sapphire/decorators';
+import { Message, MessageEmbed } from 'discord.js';
+import { DbSet, ScheduleEntity } from '#lib/database';
+
+@ApplyOptions<SkyraCommand.Options>({
+	cooldown: 5,
```
Comment (favna) [inline src/commands/Misc/Birthday/upcoming-birthdays.ts]:
```suggestion
	cooldown: 10,
```

File: src/commands/Misc/Birthday/upcoming-birthdays.ts [hunk 25]
```diff
@@ -0,0 +1,46 @@
...
+	description: LanguageKeys.Commands.Misc.UpcomingBirthdaysDescription,
+	extendedHelp: LanguageKeys.Commands.Misc.UpcomingBirthdaysExtended,
+	runIn: ['text']
+})
+export default class extends SkyraCommand {
+	public async run(message: Message, args: SkyraCommand.Args) {
```
Comment (favna) [inline src/commands/Misc/Birthday/upcoming-birthdays.ts]:
```suggestion
	public async run(message: GuildMessage, args: SkyraCommand.Args) {
```

File: src/commands/Misc/Birthday/reset-birthday.ts [hunk 26]
```diff
@@ -0,0 +1,21 @@
...
+	description: LanguageKeys.Commands.Misc.ResetBirthdayDescription,
+	extendedHelp: LanguageKeys.Commands.Misc.ResetBirthdayExtended,
+	runIn: ['text']
+})
+export default class extends SkyraCommand {
+	public async run(message: Message, args: SkyraCommand.Args) {
```
Comment (favna) [inline src/commands/Misc/Birthday/reset-birthday.ts]:
```suggestion
	public async run(message: GuildMessage, args: SkyraCommand.Args) {
```

File: src/commands/Misc/Birthday/reset-birthday.ts [hunk 27]
```diff
@@ -0,0 +1,21 @@
...
+		const birthdayReminder = this.context.client.schedules.queue.find(
+			(schedule) => schedule.taskID === 'birthday' && schedule.data.guildID === message.guild!.id && schedule.data.userID === message.author.id
```
Comment (favna) [inline src/commands/Misc/Birthday/reset-birthday.ts]:
```suggestion
			(schedule) => schedule.taskID === 'birthday' && schedule.data.guildID === message.guild.id && schedule.data.userID === message.author.id
```

File: src/commands/Misc/Birthday/upcoming-birthdays.ts [hunk 28]
```diff
@@ -0,0 +1,46 @@
...
+		const schedules = [
+			...this.context.client.schedules.queue
+				.filter((schedule) => schedule.taskID === 'birthday' && schedule.data.guildID === message.guild!.id)
```
Comment (favna) [inline src/commands/Misc/Birthday/upcoming-birthdays.ts]:
```suggestion
				.filter((schedule) => schedule.taskID === 'birthday' && schedule.data.guildID === message.guild.id)
```

File: src/commands/Misc/Birthday/upcoming-birthdays.ts [hunk 29]
```diff
@@ -0,0 +1,46 @@
...
+				.reduce((a, b) => {
+					const key = b.time.getTime();
+					if (a.has(key)) a.get(key)!.push(b);
+					else a.set(key, [b]);
+					return a;
```
Comment (favna) [inline src/commands/Misc/Birthday/upcoming-birthdays.ts]:
```suggestion
					const valueAtA = a.get(key);
					valueAtA ? valueAtA.push(b) : a.set(key, [b]);
					return a;
```

This way the map doesn't need to be accessed twice.

File: src/commands/Misc/Birthday/upcoming-birthdays.ts [hunk 30]
```diff
@@ -0,0 +1,46 @@
...
+			.setTitle(args.t(LanguageKeys.Commands.Misc.UpcomingBirthdaysTitle));
+
+		for (const [time, users] of schedules.slice(0, 10)) {
+			const birthday = new Date(Number(time));
+			embed.addField(
+				args.t(LanguageKeys.Globals.DateValue, { value: birthday }),
+				users
+					.map(
+						(schedule) =>
+							`<@${schedule.data.userID}> (${new Date().getFullYear() - new Date(schedule.data.birthDate as string).getFullYear()})`
```
Comment (favna) [inline src/commands/Misc/Birthday/upcoming-birthdays.ts:40]: Date should go through i18n so it's formatted with the server's i18n because Americans format differently from British for example.

File: src/commands/Misc/Birthday/view-birthday.ts [hunk 31]
```diff
@@ -0,0 +1,27 @@
+import { SkyraCommand } from '#lib/structures';
+import { LanguageKeys } from '#lib/i18n/languageKeys';
+import { ApplyOptions } from '@sapphire/decorators';
+import type { Message } from 'discord.js';
+
+@ApplyOptions<SkyraCommand.Options>({
+	cooldown: 5,
```
Comment (favna) [inline src/commands/Misc/Birthday/view-birthday.ts]:
```suggestion
	cooldown: 10,
```

File: src/commands/Misc/Birthday/view-birthday.ts [hunk 32]
```diff
@@ -0,0 +1,27 @@
...
+	description: LanguageKeys.Commands.Misc.ViewBirthdayDescription,
+	extendedHelp: LanguageKeys.Commands.Misc.ViewBirthdayExtended,
+	runIn: ['text']
+})
+export default class extends SkyraCommand {
+	public async run(message: Message, args: SkyraCommand.Args) {
```
Comment (favna) [inline src/commands/Misc/Birthday/view-birthday.ts]:
```suggestion
	public async run(message: GuildMessage, args: SkyraCommand.Args) {
```

File: src/commands/Misc/Birthday/view-birthday.ts [hunk 33]
```diff
@@ -0,0 +1,27 @@
...
+		const user = args.finished ? message.author! : await args.pick('userName');
+		const birthDate = this.context.client.schedules.queue.find(
+			(schedule) => schedule.taskID === 'birthday' && schedule.data.guildID === message.guild!.id && schedule.data.userID === user.id
```
Comment (favna) [inline src/commands/Misc/Birthday/view-birthday.ts]:
```suggestion
			(schedule) => schedule.taskID === 'birthday' && schedule.data.guildID === message.guild.id && schedule.data.userID === user.id
```

File: src/languages/en-US/commands/misc.json [hunk 34]
```diff
@@ -1,4 +1,5 @@
 {
+	"birthdayNotConfigured": "A birthday channel or role has not been configured in this server. Configure one using the conf command.",
```
Comment (favna) [inline src/languages/en-US/commands/misc.json]:
```suggestion
	"birthdayNotConfigured": "{{REDCROSS}} A birthday channel or role has not been configured in this server. You can configure both using the `{{prefix}}conf` command.",
```

And pass `commandContext.commandPrefix` to `{{prefix}}`

File: src/languages/en-US/commands/misc.json [hunk 35]
```diff
@@ -246,6 +247,29 @@
 			"GloriousGe0rge"
 		]
 	},
+	"resetBirthdayDescription": "Reset your birthday.",
+	"resetBirthdayNotSet": "{{REDCROSS}} You have not set your birthday in this server.",
+	"resetBirthdaySuccess": "{{GREENTICK}} I will not wish you in this server anymore.",
```
Comment (favna) [inline src/languages/en-US/commands/misc.json]: Where is `resetBirthdayExtended`?

Comment (favna) [inline src/languages/en-US/commands/misc.json]:
```suggestion
	"resetBirthdaySuccess": "{{GREENTICK}} I will not wish you a happy birthday in this server anymore.",
```

File: src/languages/en-US/commands/misc.json [hunk 36]
```diff
@@ -246,6 +247,29 @@
...
+	"setBirthdayDescription": "Set your birthday to be automatically notified.",
```
Comment (favna) [inline src/languages/en-US/commands/misc.json]:
```suggestion
	"setBirthdayDescription": "Set your birthday, and I will automatically notify the other members on that date.",
```

File: src/languages/en-US/commands/misc.json [hunk 37]
```diff
@@ -246,6 +247,29 @@
...
+	"setBirthdayExtended": {
+		"usages": ["Date"],
+		"extendedHelp": "This command needs the settings `birthday.role` or `birthday.channel` to be configured. On a user's birthday, I will give them the birthday role, and send `birthday.message` to the birthday channel.",
```
Comment (favna) [inline src/languages/en-US/commands/misc.json]:
```suggestion
		"extendedHelp": "This command needs the settings `birthday.role` or `birthday.channel` to be configured. On a user's birthday, I will give them the birthday role, and/or send `birthday.message` to the birthday channel.",
```

File: src/languages/en-US/commands/misc.json [hunk 38]
```diff
@@ -246,6 +247,29 @@
 			"GloriousGe0rge"
 		]
 	},
+	"resetBirthdayDescription": "Reset your birthday.",
+	"resetBirthdayNotSet": "{{REDCROSS}} You have not set your birthday in this server.",
+	"resetBirthdaySuccess": "{{GREENTICK}} I will not wish you in this server anymore.",
+	"setBirthdayDescription": "Set your birthday to be automatically notified.",
+	"setBirthdayExtended": {
+		"usages": ["Date"],
```
Comment (favna) [inline src/languages/en-US/commands/misc.json]:
```suggestion
		"usages": ["BirthDate"],
```

File: src/languages/en-US/commands/misc.json [hunk 39]
```diff
@@ -246,6 +247,29 @@
...
+		"extendedHelp": "This command needs the settings `birthday.role` or `birthday.channel` to be configured. On a user's birthday, I will give them the birthday role, and send `birthday.message` to the birthday channel.",
+		"explainedUsage": [
+			[
+				"BirthDate",
+				"Your birth-date, in the format YYYY/MM/DD."
```
Comment (favna) [inline src/languages/en-US/commands/misc.json]:
```suggestion
				"Your birth-date, in the format YYYY/MM/DD. (Year/Month/Day)"
```

File: src/languages/en-US/commands/misc.json [hunk 40]
```diff
@@ -246,6 +247,29 @@
...
+			]
+		],
+		"examples": [
+			"2002/01/01"
+		],
+		"reminder": "By using this command, you consent to Skyra saving your date of birth."
```
Comment (favna) [inline src/languages/en-US/commands/misc.json]:
```suggestion
		"reminder": "By using this command, you consent to it that we save your date of birth in our database."
```

File: src/languages/en-US/commands/misc.json [hunk 41]
```diff
@@ -246,6 +247,29 @@
...
+	},
+	"setBirthdayNotConfigured": "A birthday channel or role has not been configured in this server. Ask an admin to configure one using the conf command!",
```
Comment (favna) [inline src/languages/en-US/commands/misc.json]:
```suggestion
	"setBirthdayNotConfigured": "{{REDCROSS}} A birthday channel and/or role has not been configured in this server. Ask an admin to configure at least one using either the `{{prefix}}conf` command or through [the dashboard](https://skyra.pw)",
```

Same deal as before with passing in prefix.

File: src/languages/en-US/commands/misc.json [hunk 42]
```diff
@@ -246,6 +247,29 @@
...
+	"setBirthdayInvalidDate": "I'm sorry, but that doesn't seem right!",
+	"setBirthdayAlreadySet": "{{REDCROSS}} Your birthdate is already set to {{birthDate}}.",
+	"setBirthdayConfirmation": "Are you sure you want to change your birthday again?",
+	"setBirthdaySuccess": "{{GREENTICK}} I will wish you on {{nextBirthday, date}}.",
```
Comment (favna) [inline src/languages/en-US/commands/misc.json]:
```suggestion
	"setBirthdaySuccess": "{{GREENTICK}} I will wish you a happy birthday on {{nextBirthday, date}}.",
```

File: src/languages/en-US/commands/misc.json [hunk 43]
```diff
@@ -358,6 +382,20 @@
...
 	},
+	"viewBirthdayDescription": "View a user's birthday.",
+	"viewBirthdayExtended": {
+		"extendedHelp": "Birthdays are stored per-server. If you don't see yours, you probably haven't used the `set-birthday` command in this server yet!",
+		"usages": [
+			"@User"
+		],
+		"examples": [
+			"@Stitch#4094"
+		]
```
Comment (favna) [inline src/languages/en-US/commands/misc.json]:
```suggestion
		"usages": [
			"User"
		],
		"extendedHelp": "Birthdays are stored per-server. If you don't see yours, you probably haven't used the `{{prefix}}set-birthday` command in this server yet!",
		"explainedUsage": [
			[
				"User",
				"The user for whom to view the birthday"
			]
		],
		"examples": [
			"@Stitch#4094"
		]
```

Same deal as before with `{{prefix}}`. Also a full block review because overall `usages` is above `extendedHelp`

File: src/languages/en-US/commands/misc.json (hunk 15 above)
Comment (favna) [reply src/languages/en-US/commands/misc.json:420]: Why was this resolved without being answered @Stitch07 ?

File: src/languages/en-US/commands/misc.json [hunk 44]
```diff
@@ -358,6 +382,20 @@
...
+	},
+	"viewBirthdayNotSet": "{{user}} has not set a birthday yet. Use the `set-birthday` command to set your birthday.",
```
Comment (favna) [inline src/languages/en-US/commands/misc.json]:
```suggestion
	"viewBirthdayNotSet": "{{user}} has not set a birthday yet. Use the `{{prefix}}set-birthday` command to set your birthday.",
```

File: src/languages/en-US/settings.json [hunk 45]
```diff
@@ -1,4 +1,7 @@
 {
+	"birthdayChannel": "The channel where `birthday.message` will be sent to.",
+	"birthdayRole": "A role that will be given to users on their birthday (and removed in 24 hours).",
```
Comment (favna) [inline src/languages/en-US/settings.json]:
```suggestion
	"birthdayRole": "A role that will be given to users on their birthday (and removed 24 hours later).",
```

File: src/tasks/birthday.ts [hunk 46]
```diff
@@ -0,0 +1,95 @@
...
+
+	public async run(data: BirthdayTaskData): Promise<PartialResponseValue | null> {
+		const guild = this.context.client.guilds.cache.get(data.guildID);
+		if (!guild) return null;
+
+		const member = await guild.members.fetch(data.userID);
+		if (!member) return null;
+
+		const t = await guild.fetchT();
+		const [birthdayRole, birthdayChannel, birthdayMessage] = await guild.readSettings(['birthdayRole', 'birthdayChannel', 'birthdayMessage']);
```
Comment (favna) [inline src/tasks/birthday.ts]: Use GuildSettings thingy

**Review by favna — CHANGES_REQUESTED** [review]
Overall I sure hope every single error and success path is tested because it sure is quite a lot.

(and I lowkey don't want to do it myself lol)

File: src/tasks/birthday.ts (hunk 13 above)
Comment (favna) [reply src/tasks/birthday.ts:21]: What even changed in this suggestion 🤔 @kyranet @Stitch07

File: src/languages/en-US/commands/misc.json (hunk 15 above)
Comment (Stitch07) [reply src/languages/en-US/commands/misc.json:420]: It wasn't changed, just git merge messing up the formatting.

File: src/commands/Misc/Birthday/upcoming-birthdays.ts (hunk 30 above)
Comment (Stitch07) [reply src/commands/Misc/Birthday/upcoming-birthdays.ts:40]: It's not a Date, it's their age which is a number.

File: src/languages/en-US/commands/misc.json (hunk 35 above)
Comment (Stitch07) [reply src/languages/en-US/commands/misc.json]: I didn't know what to put there.

File: src/languages/en-US/commands/misc.json (hunk 43 above)
Comment (Stitch07) [reply src/languages/en-US/commands/misc.json]: FYI I don't need to add a prefix here bc the help command already does that.

**Review by favna — APPROVED** [review]
